components/scream/src/physics/shoc/tests/shoc_run_and_cmp.cpp

```diff
@@ -0,0 +1,200 @@
+#include "share/scream_session.hpp"
+#include "share/util/file_utils.hpp"
+#include "share/util/scream_utils.hpp"
+#include "share/scream_types.hpp"
```

Comment:
It seems like almost all of this is similar to p3_run_and_cmp.cpp, the only difference being a few types.

Reply:
I'll see if I can work up some relatively simple template magic that allows us to use this structure but still manages to convey what's going on.

Reply:
Hey @jgfouca and @ambrad ,

I think it's practical to create a `run_and_compare(int argc, char** argv)` function template that Does All the Things and takes `ICFactory` and `FortranData` template arguments, provided we flesh out "concepts" for those constructs. For example, a `FortranData` has to have an associated iterator (`FortranData::Iterator`, perhaps), and an `ICFactory` would have to be able to generate a list of parameter sets that `run_and_compare` uses to run several test configurations.

All of this is fine and good for C++ programmers who are familiar with templates, and maybe that's the audience who is going to be monkeying around with our `*_run_and_cmp` programs. But people who don't know what a C++ concept is might be lost without an explanation.

Do you have a preference for whether we create such a function template, vs just keeping separate code for `*_run_and_cmp`? It's not a ton of code we're talking about, and there might be enough irregularities that a template solution isn't actually as squeaky clean as we might envision.

Reply:
I'm fine with it as is. In the weeds: I think inheritance would work here; no need for templating, esp. in non-performance-critical code. That said, I noticed the overlap but did not comment b/c I'm OK with test code like a driver being a little less than perfect. I think your point about complexity vs. payoff here is spot on. In contrast (as an example of something we all agreed on), making the BFB routines common adds no complexity and is helpful for future parameterizations.